libs/qCC_io/ShpFilter.cpp

```diff
@@ -816,62 +836,49 @@ CC_FILE_ERROR LoadCloud(QFile& file,
 	}
 
 	//Points (An array of length NumPoints)
+	for (int32_t i = 0; i < numPoints; ++i)
 	{
-		for (int32_t i = 0; i < numPoints; ++i)
-		{
-			file.read(header, 16);
-			double x = qFromLittleEndianD(*reinterpret_cast<double*>(header));
-			double y = qFromLittleEndianD(*reinterpret_cast<double*>(header + 8));
-			CCVector3 P(static_cast<PointCoordinateType>(x + Pshift.x),
-						static_cast<PointCoordinateType>(y + Pshift.y),
-						0);
-			cloud->addPoint(P);
-		}
+		double  x, y;
+		shpStream >> x >> y;
+		CCVector3 P(static_cast<PointCoordinateType>(x + Pshift.x),
+					static_cast<PointCoordinateType>(y + Pshift.y),
+					0);
+		cloud->addPoint(P);
 	}
 
 	//3D clouds
 	if (isESRIShape3D(shapeType))
 	{
 		//Z boundaries
-		{
-			file.read(header, 16);
-			//DGM: ignored
-			//double zMin = qFromLittleEndianD(*reinterpret_cast<double*>(header  ));
-			//double zMax = qFromLittleEndianD(*reinterpret_cast<double*>(header+8));
-		}
+		shpStream.skipRawData(2 * sizeof(double));
 
 		//Z coordinates (an array of length NumPoints)
+		for (int32_t i = 0; i < numPoints; ++i)
 		{
-			for (int32_t i = 0; i < numPoints; ++i)
-			{
-				file.read(header, 8);
-				double z = qFromLittleEndianD(*reinterpret_cast<double*>(header));
-				const CCVector3* P = cloud->getPoint(i);
-				const_cast<CCVector3*>(P)->z = static_cast<PointCoordinateType>(z + Pshift.z);
-			}
-			cloud->invalidateBoundingBox();
+			double z;
+			shpStream >> z;
+			const CCVector3* P = cloud->getPoint(i);
+			const_cast<CCVector3*>(P)->z = static_cast<PointCoordinateType>(z + Pshift.z);
```

Comment:
@dgirardeau Is there a better way of doing this rather than this const_casting? To me const casting is usually a sign that something's not right...

Reply:
The ccPointCloud class has the 'point' method that returns a non-const pointer on a point.

Reply:
OK, thank you. I assumed it was there somewhere, but I couldn't find it because I was looking for a non-const getPoint().

Reply:
The 'point' function is protected, should it be made public ?

Reply:
Ah! Sorry I forgot that. Actually until recently it was on purpose that point couldn't be changed easily (because it invalidates the bounding box, the octree, etc.).

Then I think that we should either add a dediated method (like 'setPoint(index, P)') ... or keep the actual code as it clearly shows that this is something "dangerous" ;)

Reply:
Another solution could be to read the xs, ys , zs(if any) into a std::vector and building the cloud after instead of doing it in place like it is done now

Reply:
It could work if there are not too many points. But frankly, I think it's perfectly ok in this case to keep the const_cast and don't bother that much ;)